assignment-client/src/avatars/AvatarMixerSlavePool.cpp

```diff
@@ -151,53 +158,61 @@ void AvatarMixerSlavePool::setNumThreads(int numThreads) {
 }
 
 void AvatarMixerSlavePool::resize(int numThreads) {
-    assert(_numThreads == (int)_slaves.size());
+    assert(_data.numThreads == (int)_workers.size());
 
-    qDebug("%s: set %d threads (was %d)", __FUNCTION__, numThreads, _numThreads);
+    qDebug("%s: set %d threads (was %d)", __FUNCTION__, numThreads, _data.numThreads);
 
-    Lock lock(_mutex);
-
-    if (numThreads > _numThreads) {
+    if (numThreads > _data.numThreads) {
         // start new slaves
-        for (int i = 0; i < numThreads - _numThreads; ++i) {
-            auto slave = new AvatarMixerSlaveThread(*this, _slaveSharedData);
-            slave->start();
-            _slaves.emplace_back(slave);
+        for (int i = 0; i < numThreads - _data.numThreads; ++i) {
+            auto worker = new AvatarMixerWorkerThread(_data, _slaveSharedData);
+            worker->start();
+            _workers.emplace_back(worker);
         }
-    } else if (numThreads < _numThreads) {
-        auto extraBegin = _slaves.begin() + numThreads;
+    } else if (numThreads < _data.numThreads) {
+        auto extraBegin = _workers.begin() + numThreads;
 
         // mark slaves to stop...
-        auto slave = extraBegin;
-        while (slave != _slaves.end()) {
-            (*slave)->_stop = true;
-            ++slave;
+        auto worker = extraBegin;
+        while (worker != _workers.end()) {
+            (*worker)->stop();
+            ++worker;
         }
 
         // ...cycle them until they do stop...
-        _numStopped = 0;
-        while (_numStopped != (_numThreads - numThreads)) {
-            _numStarted = _numFinished = _numStopped;
-            _slaveCondition.notify_all();
-            _poolCondition.wait(lock, [&] {
-                assert(_numFinished <= _numThreads);
-                return _numFinished == _numThreads;
-            });
+        _data.numStopped = 0;
+        while (_data.numStopped != (_data.numThreads - numThreads)) {
+            {
+                Lock workerLock(_data.workerMutex);
+                _data.numStarted = _data.numFinished = 0;
+                _data.workerCondition.notify_all();
+            }
+            
+            {
+                Lock poolLock(_data.poolMutex);
+                if (_data.numFinished < _data.numThreads) {
+                    _data.poolCondition.wait(poolLock, [&] {
```

Comment:
hm…is it possible that all the threads will finish and notify between line 188 above and here?  so this wait would never finish?

Reply:
Heh, I thought about that this morning.  The most recent commit now has all "condition.notify" calls holding the corresponding lock when calling it (thus forcing the notify to be before line 192 or after line 194) and the "if" statement wrapping all "condition.wait" calls bypasses it if the conditions for it to be notified already exist.